AppView and Menubar (stubs) Added

chynu opened the pull request:
Created an `AppView` class (in `app_view.js`) next to `app_controller.js` to deal with all view-related functions for main application. Currently, `AppView` initializes the menubar that has the basic operations necessary for navigating the application.

These are mostly stubs -- there are no actions associated with most of the buttons, but I thought I should make a pull request so that other people can start linking actions to the menubar.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-devtools/23)
<!-- Reviewable:end -->

Commit:
Adding some dysfunctional files for devtools v2

Commit:
Somewhat functioning block name input added. Still WIP.

Commit:
minor changeS

Commit:
Added devtools

Commit:
Initial version of popup working somewhat

Commit:
Popup mostly working. Unfixed issues- configuring block name

Commit:
Popup working on load.

Commit:
Minor changes wrt popup.

chynu committed:
Merge branch 'master' into electron_test

chynu committed:
Merge pull request #1 from chynu/electron_test

Electron test merge to master

Commit:
Removed Electron modules

Commit:
new branch

Commit:
New block popup working, currently developing auto-correct

Commit:
Style changes for PR

Commit:
added to .gitignore

Commit:
removed unncessary files

Commit:
More style changes for PR

Commit:
Removed unnecessary library

Commit:
Minor stylistic changes

Commit:
Fixed some moderately major bugs.

Commit:
Minor change to test 2FA

Commit:
Responding to PR feedback (pt 1)

Commit:
Responding to more PR feedback, creating FileManagerController and View classes.

Commit:
Added new classes

Commit:
added template images, responded to PR but code is slightly broken.

Commit:
Finished responding to PR feedback.

Commit:
Minor changes to tidy code.

Commit:
Removed popup methods into a separate class completely. Responded to more PR feedback.

Commit:
Minor stylistic changes.

Commit:
removed unnecessary code

Commit:
New branch for menu navbar

Commit:
Stylistic changes for PR.

Commit:
Stylistic edits, minor functional changes.

Commit:
fixed merge conflict from feature_popup

Commit:
Added menubar, bugfixes in new block function.

Commit:
Added renamed files.

Commit:
Added renamed files, somehow got removed previously.

Commit:
Attempts at fixing new block reset bug

Commit:
Responded to PR.

Commit:
Minor edits to style (mostly just spacing)

Commit:
Responded to PR feedback

Commit:
Fix merge conflicts from feature_popup

Commit:
Minor stylistic tidying

Commit:
Merge conflicts from develop fixed

Commit:
Created AppView class, menu items in this class.

Commit:
Minor changes to addMenuAction().

AnmAtAnm commented on src/app_view.js:
```diff
@@ -0,0 +1,277 @@
+        },
+        'Open Project': null,
+        'Save': null,
+        'Save Only': {
```
Fine for now, but as I look at these two, I'm wondering if these will be dependent on the project configuration. One has to configure the locations of each platform before one can save it. A project may only have iOS configured, so does it make sense to have a save item for the others? Maybe the others are just disabled, or maybe they trigger the project configuration screen.

Also, It might read cleaner to make "Save" the menu group, and add an "All" at the top.

AnmAtAnm commented on src/app_view.js:
```diff
@@ -0,0 +1,277 @@
+          'Toolbox': null,
+          'Workspace': null
+        },
+        'Create Application': {
```
There is enough nuance to iOS and Android project setup, I don't think we want to take these on.  Let's just leave a stub for web only.

AnmAtAnm commented on src/app_view.js:
```diff
@@ -0,0 +1,277 @@
+
+    // FILE > NEW dropdown.
+    let file_new = new nw.Menu();
+    this.menuTree['File']['New']['New Project'] =
```
The duplication of the menu hierarchy here and above bothers me.  There is an opportunity for the two to get out of sync. In this first one item, you've specified 'New Project' three times: once above, once to the left of the assignment, and again as a parameter.

Consider adjusting your `addMenuAction` helper function to actually use `path` as a paththat will take in a menu path (string array), and find or generate any necessary menuTree hierarchy that doesn't already exist. Something like...

```
function addMenuAction(pathToItem, clickHandler) { ... }

addMenuAction(['File', 'New', 'Project'], () => { ... })
```

Then, there is only one place to edit it each path.

AnmAtAnm commented on begin.html:
```diff
@@ -0,0 +1,38 @@
+  <title>Blockly Demo: Blockly Developer Tools</title>
+  <script src="node_modules/jquery/dist/jquery.min.js"></script>
+  <link rel="stylesheet" href="src/factory.css">
+  <script src="https://cdn.rawgit.com/google/code-prettify/master/loader/run_prettify.js"></script>
```
I don't think this window will need prettify.

rachel-fenichel commented:
Review status: 0 of 5 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[begin.html, line 1 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-Kmbkxx71H5SOOoIobus:-Kmbkxx71H5SOOoIobut:b-7g0rcu) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/begin.html#L1)):*
> ```HTML
> <!DOCTYPE html>
> ```

Attaching screenshots can be helpful when you're creating new pages or layouts.  (I have no idea what this will look like.)

---

*[src/app_view.js, line 87 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmbleGMGpcoQoj3dIiM:-KmbleGMGpcoQoj3dIiN:b74pjop) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L87)):*
> ```JavaScript
>     let help = new nw.Menu();
> 
>     // FILE > NEW dropdown.
> ```

I would break this into helper functions by the categories that you've already called out in these comments.  There's too much going on in this function.

---

*[src/app_view.js, line 207 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmbltHs2rmiQArNikSF:-KmbltHs2rmiQArNikSG:b-64m91o) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L207)):*
> ```JavaScript
>    */
>   addMenuAction(menu, path, fcn) {
>     // TODO(celinechoo): Complete addMenuAction() to simplify adding actionable
> ```

Instead of TODO(name), do TODO(#number), and create a corresponding issue on github.  That lets you see all of your issues without reading through code, and also means someone else can solve the problem for you if they're motivated.

---

*[src/app_view.js, line 220 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmbmH1DVKW_7PiH0M9x:-KmbmH1DVKW_7PiH0M9y:b1n99mb) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L220)):*
> ```JavaScript
>    * Helper function to add menu items to the main menubar.
>    *
>    * @param {!Menu} menu Menu object from NW.js to which we are adding an element.
> ```

Is the full type just `Menu`?  Usually it's part of a namespace (e.g. `Blockly.Menu`).

---

*[src/app_view.js, line 223 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmbmPm2WxGgVrHhafN9:-KmbmPm2WxGgVrHhafNA:bj3j3y3) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L223)):*
> ```JavaScript
>    * @param {string} name Name of element being added to menu.
>    * @param {!Menu} subMenu Menu object to be added if the added element will also
>    * have its own sub-dropdown. Optional (if unused, will be null).
> ```

Nit: four more spaces at the beginning of this line, and for two lines below.

Lines that continue an annotation always have a four-space additional indent.

---

*[src/app_view.js, line 259 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmbmyeAcjzID1335Epf:-KmbmyeAcjzID1335Epg:b-5q1v8c) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L259)):*
> ```JavaScript
>         menuItem = menuItem[pathToMenuItem[i]];
>       } catch(e) {
>         console.log(e);
> ```

Will there be enough information in the log for you to debug?  If not, add a message with some context.

Is this a recoverable error?  Or does it break the whole application?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/23)*
<!-- Sent from Reviewable.io -->

chynu committed:
Simplified menuTree init.

chynu committed:
Simplified menuTree init

rachel-fenichel commented:
lgtm overall.  Make sure you squash when merging.

---

Reviewed 2 of 11 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[src/app_view.js, line 30 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmcqOQGShOCTEYgKO4_:-KmcqOQGShOCTEYgKO4a:bl9h003) ([raw file](https://github.com/google/blockly-devtools/blob/18ed30fab7429532089950b2e741416aa406f110/src/app_view.js#L30)):*
> ```JavaScript
> 
> /**
>  * @fileoverview The AppView Class deals with the visual parts of the main
> ```

Nit: fileoverview comes before the requires: https://github.com/google/blockly-devtools/blob/master/src/block_library_storage.js#L21

The comment here would describe the class.  Sometimes that will be similar to the fileoverview text, but not always.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/23#-:-Kmcq_5ZSoyZmnVtjBcJ:bfu03fk)*
<!-- Sent from Reviewable.io -->

chynu committed:
Minor changes

chynu commented:
Review status: 3 of 4 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/app_view.js, line 51 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmZELHbKE81qZGmIxj6-r3-51:-Kmcpbgy3eGS3E5uS_Xj:brfgse7) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L51)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Fine for now, but as I look at these two, I'm wondering if these will be dependent on the project configuration. One has to configure the locations of each platform before one can save it. A project may only have iOS configured, so does it make sense to have a save item for the others? Maybe the others are just disabled, or maybe they trigger the project configuration screen.

Also, It might read cleaner to make "Save" the menu group, and add an "All" at the top.
</blockquote></details>

I consolidated Save into one submenu, and left the four options for now.

---

*[src/app_view.js, line 62 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmZELHbKE81qZGmIxj6-r3-62:-KmcpZs4jBlGKerfqxfC:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L62)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

There is enough nuance to iOS and Android project setup, I don't think we want to take these on.  Let's just leave a stub for web only.
</blockquote></details>

Done.

---

*[src/app_view.js, line 87 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmbleGMGpcoQoj3dIiM:-KmcphPUgPyOwMQvfrWY:bs1v41b) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L87)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

I would break this into helper functions by the categories that you've already called out in these comments.  There's too much going on in this function.
</blockquote></details>

Done! Created a simpler menubar init function.

---

*[src/app_view.js, line 89 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmZELHbKE81qZGmIxj6-r3-89:-KmcpuKiqLfo7V1fSbRX:bc6vjbw) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L89)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

The duplication of the menu hierarchy here and above bothers me.  There is an opportunity for the two to get out of sync. In this first one item, you've specified 'New Project' three times: once above, once to the left of the assignment, and again as a parameter.

Consider adjusting your `addMenuAction` helper function to actually use `path` as a paththat will take in a menu path (string array), and find or generate any necessary menuTree hierarchy that doesn't already exist. Something like...

```
function addMenuAction(pathToItem, clickHandler) { ... }

addMenuAction(['File', 'New', 'Project'], () => { ... })
```

Then, there is only one place to edit it each path.
</blockquote></details>

I removed the addMenuAction function because I realized my implementation of the menubar init was very disorganized.

---

*[src/app_view.js, line 207 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmbltHs2rmiQArNikSF:-KmcqnXjac7E0KxrH2Vc:b-eeywll) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L207)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Instead of TODO(name), do TODO(#number), and create a corresponding issue on github.  That lets you see all of your issues without reading through code, and also means someone else can solve the problem for you if they're motivated.
</blockquote></details>

Noted, thanks!

---

*[src/app_view.js, line 220 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmbmH1DVKW_7PiH0M9x:-Kmcqp9PNeaUM2bLzZ-0:bs7wumx) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L220)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Is the full type just `Menu`?  Usually it's part of a namespace (e.g. `Blockly.Menu`).
</blockquote></details>

Done. 

---

*[src/app_view.js, line 223 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmbmPm2WxGgVrHhafN9:-Kmcqq9d_J2Thh0QRy6O:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L223)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Nit: four more spaces at the beginning of this line, and for two lines below.

Lines that continue an annotation always have a four-space additional indent.
</blockquote></details>

Done.

---

*[src/app_view.js, line 259 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmbmyeAcjzID1335Epf:-Kmcquga2zNENJ1r5rCh:b4bt2zg) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/src/app_view.js#L259)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Will there be enough information in the log for you to debug?  If not, add a message with some context.

Is this a recoverable error?  Or does it break the whole application?
</blockquote></details>

Removed these lines, but will note the next time I use try/catch.

---

*[src/app_view.js, line 30 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmcqOQGShOCTEYgKO4_:-Kmcr4fsX5P0w_aV6zp_:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/18ed30fab7429532089950b2e741416aa406f110/src/app_view.js#L30)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Nit: fileoverview comes before the requires: https://github.com/google/blockly-devtools/blob/master/src/block_library_storage.js#L21

The comment here would describe the class.  Sometimes that will be similar to the fileoverview text, but not always.
</blockquote></details>

Done.

---

*[begin.html, line 1 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-Kmbkxx71H5SOOoIobus:-KmbnhQ6pXz6ERrBmQQ1:b-z94cyw) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/begin.html#L1)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Attaching screenshots can be helpful when you're creating new pages or layouts.  (I have no idea what this will look like.)
</blockquote></details>

I didn't mean to add this page as part of this pull request. Bad oversight on my part.

---

*[begin.html, line 9 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/23#-KmZELHbKE81qZGmIxj5-r3-9:-KmbnssU7fWI3RfJ5Qq5:b7e96us) ([raw file](https://github.com/google/blockly-devtools/blob/06453b53371027437457cffbb94f5c461ad6c947/begin.html#L9)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I don't think this window will need prettify.
</blockquote></details>

Bad oversight on my part -- this shouldn't be part of this pull request.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/23)*
<!-- Sent from Reviewable.io -->

chynu committed:
Added AppView
